developer/js/source/errors.ts

```diff
@@ -0,0 +1,171 @@
+/**
+ * Log levels.
+ * 
+ * Note: Currently, this acts like a bit set, where the upper 4 bits of an
+ * unsigned 16 bit value are the log level flags.
+ */
+export enum LogLevel {
+  CERR_FATAL = 0x8000,
+  CERR_ERROR = 0x4000,
+  CERR_WARNING = 0x2000,
+  CERR_MEMORY = 0x1000,  // N.B., probably unused in TypeScript
+  CERR_INFO = 0x0000,  // N.B., not in widespread use
+};
+
+/**
+ * Error codes. Use these when logging messages.
+ * 
+ * Extends https://github.com/keymanapp/keyman/blob/99db3c0d2448f448242e6397f9d72e9a7ccee4b9/windows/src/global/inc/Comperr.h
+ */
+export enum KeymanCompilerError {
+  CERR_LEXICAL_MODEL_MIN = 0x0800,
+  CERR_LEXICAL_MODEL_MAX = 0x08FF,
+
+  CERR_FATAL_LM = LogLevel.CERR_FATAL | CERR_LEXICAL_MODEL_MIN,
+  /* Place all fatal LM compiler errors here! */
+
+  CERR_ERROR_LM = LogLevel.CERR_ERROR | CERR_LEXICAL_MODEL_MIN,
+  /* Place all recoverable LM compiler errors here! */
+
+  CERR_WARN_LM = LogLevel.CERR_WARNING | CERR_LEXICAL_MODEL_MIN,
+  /* Place all LM compiler warnings here! */
+  MixedNormalizationForms,
+  DuplicateWordInSameFile,
+}
+
+/**
+ * Human-readable titles for the various log levels.
+ * 
+ * Taken from https://github.com/keymanapp/keyman/blob/d83cfffe511ce65b781f919e89e3693146844849/windows/src/developer/TIKE/project/Keyman.Developer.System.Project.ProjectLog.pas#L39-L46
+ */
+const LOG_LEVEL_TITLE: {[level in LogLevel]: string} = {
+  [LogLevel.CERR_INFO]: '',
+  [LogLevel.CERR_WARNING]: 'Warning',
+  [LogLevel.CERR_ERROR]: 'Error',
+  [LogLevel.CERR_FATAL]: 'Fatal Error',
+  [LogLevel.CERR_MEMORY]: '???',
```

Comment:
"Out of memory" -- kinda unlikely to arise these days. When the Keyman .kmx compiler was first written, it was a distinct possibility; my computer had 4MB RAM. 👴